### PR TITLE
Analyzer: Implement support for the Bower package manager

### LINKED ISSUE #89: Add Bower (JavaScript) as supported package manager
Add support for the Bower package manager to the analyzer module.

https://bower.io/

### PR DESCRIPTION
**Pull request by pnowakowski**
Resolving Bower dependencies involves the following steps:
- move `bower_components` folder to temp location if it exists
- install bower dependencies with `bower install`
- parse dependencies from `bower list --json` (for each level of
dependencies, value of `dependencies` key provide details of dependant
packages, that again hold details for it's own dependencies)
- remove `bower_components` that was created with `bower install`
- restore original `bower_components` from temporary folder (if exists)

Resolves #89.

---
This change is [Reviewable](https://reviewable.io/reviews/heremaps/oss-review-toolkit/282)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -41,6 +42,7 @@ install:
   - wget -q https://dl.google.com/android/repository/sdk-tools-linux-3859397.zip
   - unzip -q sdk-tools-linux-3859397.zip -d $ANDROID_HOME
   - yes | $ANDROID_HOME/tools/bin/sdkmanager --verbose "platform-tools"
+  - gem install licensee -v 9.8.0
```
Comment (mnonnenmacher) [inline .travis.yml]: Is this related to this change?

File: analyzer/src/funTest/assets/projects/synthetic/bower-expected-output.yml [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+      type: ""
+      url: ""
+      revision: ""
+      path: ""
+    vcs_processed:
+      type: ""
+      url: ""
+      revision: ""
+      path: ""
+  curations: []
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/assets/projects/synthetic/bower-expected-output.yml]: The `Project` itself should not appear in the package list.

File: analyzer/src/funTest/assets/projects/synthetic/bower/bower.json [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  "name": "a-test-project",
+  "authors": [
+    "author-xyz"
+  ],
+  "description": "This string describes the test project.",
+  "main": "",
+  "keywords": [
+    "keyword-test"
+  ],
+  "license": "MIT",
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/assets/projects/synthetic/bower/bower.json]: License should be `Apache-2.0`.

File: analyzer/src/funTest/kotlin/BowerTest.kt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+package com.here.ort.analyzer
+
+import com.here.ort.analyzer.managers.Bower
+import com.here.ort.downloader.VersionControlSystem
+import com.here.ort.model.yamlMapper
+import com.here.ort.utils.normalizeVcsUrl
+import com.here.ort.utils.test.DEFAULT_ANALYZER_CONFIGURATION
+import com.here.ort.utils.test.DEFAULT_REPOSITORY_CONFIGURATION
+import com.here.ort.utils.test.USER_DIR
+import com.here.ort.utils.test.patchExpectedResult
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/BowerTest.kt]: Please add an empty line after this.

File: analyzer/src/funTest/kotlin/BowerTest.kt [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import io.kotlintest.matchers.beEmpty
+import io.kotlintest.should
+import io.kotlintest.shouldBe
+import io.kotlintest.shouldNotBe
+import io.kotlintest.specs.StringSpec
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/BowerTest.kt]: Please add an empty line after this.

File: analyzer/src/funTest/kotlin/integration/PolymerIntegrationTest.kt [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+package com.here.ort.analyzer.integration
+
+import com.here.ort.analyzer.PackageManagerFactory
+import com.here.ort.analyzer.managers.Bower
+import com.here.ort.analyzer.managers.NPM
+import com.here.ort.model.Identifier
+import com.here.ort.model.Package
+import com.here.ort.model.RemoteArtifact
+import com.here.ort.model.VcsInfo
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/integration/PolymerIntegrationTest.kt]: Please add an empty line after this.

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 7]
```diff
@@ -19,20 +19,265 @@
...
+import com.here.ort.model.PackageReference
+import com.here.ort.model.Project
+import com.here.ort.model.ProjectAnalyzerResult
+import com.here.ort.model.RemoteArtifact
+import com.here.ort.model.Scope
+import com.here.ort.model.VcsInfo
 import com.here.ort.model.config.AnalyzerConfiguration
 import com.here.ort.model.config.RepositoryConfiguration
+import com.here.ort.model.jsonMapper
+
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Bower.kt]: Please remove the empty line.

File: analyzer/src/funTest/kotlin/integration/PolymerIntegrationTest.kt [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+import java.io.File
+
+class PolymerIntegrationTest : AbstractIntegrationSpec() {
+
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/integration/PolymerIntegrationTest.kt]: Please remove the empty line to align code style with the other classes.

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 9]
```diff
@@ -19,20 +19,265 @@
...
         override val globsForDefinitionFiles = listOf("bower.json")
 
         override fun create(analyzerConfig: AnalyzerConfiguration, repoConfig: RepositoryConfiguration) =
                 Bower(analyzerConfig, repoConfig)
     }
+
+    companion object {
+        /* We do not actually depend on any features specific to Bower version 1.8.2, but we still want to
+          stick to fixed versions to be sure to get consistent results. */
+        private const val REQUIRED_BOWER_VERSION = "1.8.2"
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Bower.kt]: Should we allow at least `1.8.*`? You could check `NPM.kt` for an example for version ranges.

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 10]
```diff
@@ -19,20 +19,265 @@
...
+
+        private fun extractDeclaredLicenses(node: JsonNode): SortedSet<String> {
+            val result = mutableSetOf<String>()
+
+            val license = node["pkgMeta"]["license"].textValueOrEmpty()
+            if (!license.isEmpty()) {
+                result.add(license)
+            }
+
+            return result.toSortedSet()
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Bower.kt]: You could use this to avoid needing `val result`:

```kotlin
return if (license.isNotEmpty()) {
    sortedSetOf(license)
} else {
    sortedSetOf<String>() // Generic type probably not required.
}
```

Comment (mnonnenmacher) [reply analyzer/src/main/kotlin/managers/Bower.kt]: Another way is:

```kotlin
return sortedSetOf<String>().apply {
    if (license.isNotEmpty()) {
        add(license)
    }
}
```

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 11]
```diff
@@ -19,20 +19,265 @@
...
+        private fun extractPackages(node: JsonNode): Map<String, Package> {
+            val result = mutableMapOf<String, Package>()
+
+            val stack = Stack<JsonNode>()
+            stack.push(node)
+
+            while (!stack.empty()) {
+                val currentNode = stack.pop()
+                val pkg = extractPackage(currentNode)
+                result.put("${pkg.id.name}:${pkg.id.version}", pkg)
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Bower.kt]: `result["${pkg.id.name}:${pkg.id.version}"] = pkg`

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 12]
```diff
@@ -19,20 +19,265 @@
...
+
+                val children = currentNode["dependencies"].fields().asSequence().map { it.value }
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Bower.kt]: I would assume `asSequence()` is not required here.

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 13]
```diff
@@ -19,20 +19,265 @@
...
+        jsonMapper.enable(SerializationFeature.INDENT_OUTPUT)
+        checkVersion(
+                Requirement.buildNPM(REQUIRED_BOWER_VERSION),
+                ignoreActualVersion = analyzerConfig.ignoreToolVersions
+        )
+        return definitionFiles
+    }
+
+    override fun resolveDependencies(definitionFile: File): ProjectAnalyzerResult? {
+        log.info( "Resolving dependencies for: $definitionFile" )
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Bower.kt]: Use this syntax to avoid building the String if `info` is not loggable:

```kotlin
log.info { "..." }
```

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 14]
```diff
@@ -19,20 +19,265 @@
...
+        }
+
+        return result
+    }
+
+    private fun stashDir(originalDir: File) {
+        if (!originalDir.isDirectory) return
+
+        val tempDir = createTempDir("analyzer", ".tmp", originalDir)
+        log.info("Temporarily moving directory from '$originalDir' to '$tempDir'.")
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Bower.kt]: `log.info { ... }`

File: utils/src/main/kotlin/Utils.kt [hunk 15]
```diff
@@ -391,6 +392,12 @@ fun JsonNode?.textValueOrEmpty(): String = this?.textValue()?.let { it } ?: ""
  */
 fun String.encodeOrUnknown() = fileSystemEncode().takeUnless { it.isBlank() } ?: "unknown"
 
+/*
+ * Convenience function for [JsonNode] that returns the an empty iterator
```
Comment (mnonnenmacher) [inline utils/src/main/kotlin/Utils.kt]: Remove "the", also make use of the 120 chars line length limit.

File: utils/src/main/kotlin/Utils.kt [hunk 16]
```diff
@@ -391,6 +392,12 @@ fun JsonNode?.textValueOrEmpty(): String = this?.textValue()?.let { it } ?: ""
...
+ * if [JsonNode.fieldNames] is called on a null object or the field names otherwise.
```
Comment (mnonnenmacher) [inline utils/src/main/kotlin/Utils.kt]: Add a comma before "or" for better readability.

File: .travis.yml (hunk 1 above)
Comment (fviernau) [reply .travis.yml]: Done.

File: analyzer/src/funTest/assets/projects/synthetic/bower-expected-output.yml (hunk 2 above)
Comment (fviernau) [reply analyzer/src/funTest/assets/projects/synthetic/bower-expected-output.yml]: Done.

File: analyzer/src/funTest/assets/projects/synthetic/bower/bower.json (hunk 3 above)
Comment (fviernau) [reply analyzer/src/funTest/assets/projects/synthetic/bower/bower.json]: Done.

File: analyzer/src/funTest/kotlin/BowerTest.kt (hunk 4 above)
Comment (fviernau) [reply analyzer/src/funTest/kotlin/BowerTest.kt]: Done.

File: analyzer/src/funTest/kotlin/BowerTest.kt (hunk 5 above)
Comment (fviernau) [reply analyzer/src/funTest/kotlin/BowerTest.kt]: Done.

File: analyzer/src/funTest/kotlin/integration/PolymerIntegrationTest.kt (hunk 6 above)
Comment (fviernau) [reply analyzer/src/funTest/kotlin/integration/PolymerIntegrationTest.kt]: Done.

File: analyzer/src/main/kotlin/managers/Bower.kt (hunk 7 above)
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: Done.

File: analyzer/src/funTest/kotlin/integration/PolymerIntegrationTest.kt (hunk 8 above)
Comment (fviernau) [reply analyzer/src/funTest/kotlin/integration/PolymerIntegrationTest.kt]: Done.

File: analyzer/src/main/kotlin/managers/Bower.kt (hunk 9 above)
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: Will do in a commit on top along with some testing.

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 17]
```diff
@@ -19,20 +19,265 @@
...
+        private fun extractRepositoryType(node: JsonNode): String {
+            var result = ""
+
+            val repositoryNode = node["pkgMeta"]["repository"]
+            if ( repositoryNode != null ) {
+                result = repositoryNode["type"].textValueOrEmpty()
+            }
+
+            return result
+        }
```
[the remark this answers is not in the document]
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: nice, done.

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 18]
```diff
@@ -19,20 +19,265 @@
...
+            if (repositoryNode != null) {
+                result = repositoryNode["url"].textValueOrEmpty()
+            }
+
+            if (result.isEmpty()) {
+                result = node["pkgMeta"]["_source"].textValueOrEmpty()
+            }
+
+            return result
+        }
```
[the remark this answers is not in the document]
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: Done.

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 19]
```diff
@@ -19,20 +19,265 @@
...
+            }
+
+            var result = resolutionNode["commit"].textValueOrEmpty()
+
+            if (result.isEmpty()) {
+                result = resolutionNode["tag"].textValueOrEmpty()
+            }
+
+            return result
+        }
```
[the remark this answers is not in the document]
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: Done.

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 20]
```diff
@@ -19,20 +19,265 @@
...
+
+        private fun extractVcsInfo(node: JsonNode): VcsInfo {
+            return VcsInfo(
+                    type = extractRepositoryType(node),
+                    url = extractRepositoryUrl(node),
+                    revision = extractRevision(node)
+            )
+        }
```
[the remark this answers is not in the document]
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: Done.

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 21]
```diff
@@ -19,20 +19,265 @@
...
+                    id = extractPackageId(node),
+                    declaredLicenses = extractDeclaredLicenses(node),
+                    description = node["pkgMeta"]["description"].textValueOrEmpty(),
+                    homepageUrl = node["pkgMeta"]["homepage"].textValueOrEmpty(),
+                    binaryArtifact = RemoteArtifact.EMPTY,
+                    sourceArtifact = RemoteArtifact.EMPTY, // TODO: implement me!
+                    vcs = vcsInfo,
+                    vcsProcessed = vcsInfo.normalize()
+            )
+        }
```
[the remark this answers is not in the document]
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: Left as is as identation becomes quite big and brings other readability issues.

File: analyzer/src/main/kotlin/managers/Bower.kt (hunk 12 above)
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: field().map doesn t work

File: analyzer/src/main/kotlin/managers/Bower.kt (hunk 13 above)
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: Done.

File: utils/src/main/kotlin/Utils.kt (hunk 15 above)
Comment (fviernau) [reply utils/src/main/kotlin/Utils.kt]: Done.

File: utils/src/main/kotlin/Utils.kt (hunk 16 above)
Comment (fviernau) [reply utils/src/main/kotlin/Utils.kt]: Done.

File: analyzer/src/main/kotlin/managers/Bower.kt (hunk 14 above)
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: Done.

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 22]
```diff
@@ -19,20 +19,237 @@
...
+                result.add(packageRefence)
+            }
+
+            return result.toSortedSet()
+        }
+    }
+
+    private val stashedDirectories = HashMap<File, File>()
+
+    override fun toString(): String = PROVIDER_NAME
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/managers/Bower.kt]: Is it necessary to provide the return type here?

File: .appveyor.yml [hunk 23]
```diff
@@ -23,6 +23,7 @@ clone_depth: 50
 install:
   - git submodule update --init --recursive
   - npm install -g npm@6.4.0 yarn@1.9.4
+  - npm install -g bower@1.8.2
```
Comment (sschuberth) [inline .appveyor.yml]: Simply add this to the previous line (as a first item, sorted alphabetically), so `npm install` is only called once.

File: .travis.yml [hunk 24]
```diff
@@ -31,6 +31,7 @@ install:
   - eval "$(gimme)"
   - curl https://raw.githubusercontent.com/golang/dep/7d5cd199ce454707f81c63b7ea4299151b8b981d/install.sh | sh
   - npm install -g npm@6.4.0 yarn@1.9.4
+  - npm install -g bower@1.8.2
```
Comment (sschuberth) [inline .travis.yml]: Simply add this to the previous line (as a first item, sorted alphabetically), so `npm install` is only called once.

File: analyzer/src/funTest/kotlin/integration/PolymerIntegrationTest.kt [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+            binaryArtifact = RemoteArtifact.EMPTY,
+            sourceArtifact = RemoteArtifact.EMPTY,
+            vcs = VcsInfo(
+                    type = "Git",
+                    url = "https://github.com/Polymer/polymer.git",
+                    revision = "v2.4.0"
+            )
+    )
+
+    fun findDownloadedFiles( vararg filenames: String ): MutableList<File>
```
Comment (sschuberth) [inline analyzer/src/funTest/kotlin/integration/PolymerIntegrationTest.kt]: Please write as

    private fun findDownloadedFiles(vararg filenames: String) =
            downloadResult.downloadDirectory.walkTopDown().filter { it.name in filenames }.toList()

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 26]
```diff
@@ -19,20 +19,237 @@
 
 package com.here.ort.analyzer.managers
 
-import com.here.ort.analyzer.PackageManager
+import ch.frankel.slf4k.info
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Bower.kt]: We have a policy to always use a wildcard import for this even if only a single log level is used:

    import ch.frankel.slf4k.*

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 27]
```diff
@@ -19,20 +19,237 @@
...
+import java.util.SortedSet
+import java.util.Stack
 
 /**
  * The Bower package manager for JavaScript, see https://bower.io/.
  */
 class Bower(analyzerConfig: AnalyzerConfiguration, repoConfig: RepositoryConfiguration) :
-        PackageManager(analyzerConfig, repoConfig) {
+        PackageManager(analyzerConfig, repoConfig), CommandLineTool {
+
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Bower.kt]: Nit: Superfluous blank line.

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 28]
```diff
@@ -19,20 +19,237 @@
...
     class Factory : AbstractPackageManagerFactory<Bower>() {
         override val globsForDefinitionFiles = listOf("bower.json")
 
         override fun create(analyzerConfig: AnalyzerConfiguration, repoConfig: RepositoryConfiguration) =
                 Bower(analyzerConfig, repoConfig)
     }
+
+    companion object {
+        /* We do not actually depend on any features specific to Bower version 1.8.2, but we still want to
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Bower.kt]: Nit: Please prefix each comment line with `//` instead of using such block comments. We only use block comments for KDoc: `/** ... */`.

File: analyzer/src/main/kotlin/managers/Bower.kt (hunk 22 above)
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Bower.kt]: No, it's not 😉

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 29]
```diff
@@ -19,20 +19,237 @@
...
+        )
+        return definitionFiles
+    }
+
+    override fun resolveDependencies(definitionFile: File): ProjectAnalyzerResult? {
+        log.info { "Resolving dependencies for: '${definitionFile.absolutePath}'" }
+
+        val workingDir = definitionFile.parentFile
+        val bowerComponentsDir = File(workingDir, "bower_components")
+        var result: ProjectAnalyzerResult?
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Bower.kt]: This could be a `val`.

File: analyzer/src/main/kotlin/managers/Bower.kt [hunk 30]
```diff
@@ -19,20 +19,237 @@
...
+    }
+
+    private fun stashDir(originalDir: File) {
+        if (!originalDir.isDirectory) return
+
+        val tempDir = createTempDir("analyzer", ".tmp", originalDir)
+        log.info { "Temporarily moving directory from '${originalDir.absolutePath}' to '${tempDir.absolutePath}'." }
+
+        Files.move(originalDir.toPath(), tempDir.toPath(), StandardCopyOption.ATOMIC_MOVE)
+        stashedDirectories.put(originalDir, tempDir)
```
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Bower.kt]: This `put` can be converted to an assignment.

File: utils/src/main/kotlin/Utils.kt [hunk 31]
```diff
@@ -381,7 +382,7 @@ fun File.toSafeURI(): URI {
 }
 
 /**
- * Convenience function for [JsonNode] that returns an empty string if [JsonNode.textValue] is called on a null object
+ * Convenience function for [JsonNode] that returns an empty string if [JsonNode.textValue] is called on a null object,
```
Comment (sschuberth) [inline utils/src/main/kotlin/Utils.kt]: Nit: Ideally this would have been a separate commit.

File: utils/src/main/kotlin/Utils.kt [hunk 32]
```diff
@@ -391,6 +392,18 @@ fun JsonNode?.textValueOrEmpty(): String = this?.textValue()?.let { it } ?: ""
  */
 fun String.encodeOrUnknown() = fileSystemEncode().takeUnless { it.isBlank() } ?: "unknown"
 
+/*
```
Comment (sschuberth) [inline utils/src/main/kotlin/Utils.kt]: Please sort this in alphabetically to the other `JsonNode?` extension functions.

File: analyzer/src/main/kotlin/managers/Bower.kt (hunk 10 above)
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: Done.

File: .appveyor.yml (hunk 23 above)
Comment (fviernau) [reply .appveyor.yml]: Done.

File: analyzer/src/main/kotlin/managers/Bower.kt (hunk 27 above)
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: Done.

File: analyzer/src/main/kotlin/managers/Bower.kt (hunk 29 above)
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: That's interresting :-) Done.

File: analyzer/src/main/kotlin/managers/Bower.kt (hunk 30 above)
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: Done.

File: analyzer/src/main/kotlin/managers/Bower.kt (hunk 28 above)
Comment (fviernau) [reply analyzer/src/main/kotlin/managers/Bower.kt]: Done.

File: utils/src/main/kotlin/Utils.kt (hunk 31 above)
Comment (fviernau) [reply utils/src/main/kotlin/Utils.kt]: Done.

File: utils/src/main/kotlin/Utils.kt (hunk 32 above)
Comment (fviernau) [reply utils/src/main/kotlin/Utils.kt]: Done.

File: analyzer/src/funTest/kotlin/integration/PolymerIntegrationTest.kt (hunk 25 above)
Comment (fviernau) [reply analyzer/src/funTest/kotlin/integration/PolymerIntegrationTest.kt]: Done.

File: .appveyor.yml [hunk 33]
```diff
@@ -22,7 +22,7 @@ clone_depth: 50
 
 install:
   - git submodule update --init --recursive
-  - npm install -g npm@6.4.0 yarn@1.9.4
+  - npm install -g bower@1.8.2 npm@6.4.0 yarn@1.9.4
```
[the remark this answers is not in the document]
Comment (fviernau) [reply .appveyor.yml]: It was because I used that during implementation, done.

**Comment by sschuberth** [comment]
Merging despite the unrelated AppVeyor failure.
